src/persist/src/operators/stream.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A Timely Dataflow operator that passes through its input after persisting
+//! it.
+
+use std::convert::identity;
+use std::sync::mpsc;
+
+use timely::dataflow::operators::{Concat, OkErr};
+use timely::dataflow::{Scope, Stream};
+use timely::Data;
+
+use crate::error::Error;
+use crate::indexed::runtime::{StreamReadHandle, StreamWriteHandle};
+use crate::operators;
+use crate::Token;
+
+/// A Timely Dataflow operator that passes through its input after persisting
+/// it.
+pub trait PersistentStream<G: Scope<Timestamp = u64>, K: Data, V: Data> {
+    /// Persist each element of the stream, emitting it after it's been durably
+    /// recorded.
+    fn persisted_stream(
+        &self,
+        token: Token<StreamWriteHandle<K, V>, StreamReadHandle<K, V>>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    );
+}
+
+impl<G, K, V> PersistentStream<G, K, V> for Stream<G, ((K, V), u64, isize)>
```

Comment:
I don't feel strongly about this distinction but in my head things that enforce a `(data, time, diff)` structure on the data type of a timely `Stream` are usually `Collection`s and things that do not enforce that structure on the data are raw`Stream`s. Do you think this should be an operator on collections instead?

Reply:
So in the process of writing #7356 I realized that we're always talking about collection semantics, not stream, so I'm in the process of convincing myself that we should switch them all over. Since we use stream everywhere else at the moment, I'm inclined to leave this on as-is and fix it later in batch if we do that.